src/core/models/dt_linearmodel.cc

```diff
@@ -0,0 +1,1096 @@
+  T loss_old = T(0); // Value of `loss` for a previous iteraction
+  std::vector<T> loss_history;
+  const Column& target_col0_val = validation? dt_y_val->get_column(0)
+                                            : target_col0_train;  // whatever
+  colvec cols_val;
+
+  if (validation) {
+    cols_val = make_casted_columns<T>(dt_X_val, stype);
+    iteration_nrows = static_cast<size_t>(std::ceil(nepochs_val * static_cast<T>(iteration_nrows)));
+    niterations = total_nrows / iteration_nrows + (total_nrows % iteration_nrows > 0);
+    loss_history.resize(val_niters, 0.0);
+  }
+
+
+  // Mutex to update global feature importance information
+  std::mutex m;
+
+  // Calculate work amounts for full fit iterations, last fit iteration and
+  // validation.
+  size_t work_total = (niterations - 1) * get_work_amount(iteration_nrows);
+  work_total += get_work_amount(total_nrows - (niterations - 1) * iteration_nrows);
+  if (validation) work_total += niterations * get_work_amount(dt_X_val->nrows());
+
+  // Set work amount to be reported by the zero thread.
+  dt::progress::work job(work_total);
+  job.set_message("Fitting...");
+  NThreads nthreads = nthreads_from_niters(iteration_nrows, MIN_ROWS_PER_THREAD);
+
+  dt::parallel_region(nthreads,
+    [&]() {
+      // Each thread gets a private storage observations and feature importances.
+      tptr<T> x = tptr<T>(new T[nfeatures]);
+      tptr<T> fi = tptr<T>(new T[nfeatures]());
+
+      for (size_t iter = 0; iter < niterations; ++iter) {
+        size_t iteration_start = iter * iteration_nrows;
+        iteration_end = (iter == niterations - 1)? total_nrows :
+                                                   (iter + 1) * iteration_nrows;
+        size_t iteration_size = iteration_end - iteration_start;
+
+        // Training.
+        dt::nested_for_static(iteration_size, ChunkSize(MIN_ROWS_PER_THREAD), [&](size_t i) {
+          size_t ii = (iteration_start + i) % dt_X_train->nrows();
+          U target;
+          bool isvalid = target_col0_train.get_element(ii, &target);
+
+          if (isvalid && _isfinite(target) && read_row(ii, cols, x)) {
+            // Loop over all the labels
+            for (size_t k = 0; k < label_ids_train.size(); ++k) {
+              // Update all the weights with SGD
+              for (size_t j = 0; j < nfeatures + 1; ++j) {
+
+                T grad = linkfn(predict_row(x, k,
+                           [&](size_t f_id, T f_imp) {
+                             fi[f_id] += f_imp;
+                            }
+                         ));
+                T y = static_cast<T>(targetfn(target, label_ids_train[k]));
+                grad = 2 * dlinkfn(grad) * (grad - y);
+                if (j) grad *= x[j - 1];
+                grad += copysign(lambda1, w[k][j]); // L1 regularization
+                grad += 2 * lambda2 * w[k][j];      // L2 regularization
+                w[k][j] -= eta * grad;
```

Comment:
Are you updating the model's coefficients on the go? This is definitely not the standard way of computing regressions, so a link to a paper would be necessary. I saw https://www.jmlr.org/papers/volume15/dhurandhar14a/dhurandhar14a.pdf, but that's probably not what you were implementing.

Reply:
Isn't it a standard SGD method? https://en.wikipedia.org/wiki/Stochastic_gradient_descent#Iterative_method

Reply:
Ah yes, I didn't recognize it. Is it really better in the context of a linear regression?